### PR TITLE
image-discovery: add per-user scope

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/shared/discover-image.c [hunk 1]
```diff
@@ -548,12 +550,95 @@ static int image_make(
...
+                r = pick_image_search_path(RUNTIME_SCOPE_SYSTEM, class, &b);
+                if (r < 0)
+                        return r;
+
+                r = strv_extend_strv(&a, b, /* filter_duplicates= */ false);
+                if (r < 0)
+                        return r;
+
+                *ret = TAKE_PTR(a);
+                return r;
```
Comment (bluca) [inline src/shared/discover-image.c]: `return 0`

**Review by bluca — CHANGES_REQUESTED** [review]
Would it be possible to have some validation for the nspawn side of things?

Also please add the dissect options to shell completion, all that is needed is the following diff:

```
--- a/shell-completion/bash/systemd-dissect
+++ b/shell-completion/bash/systemd-dissect
@@ -29,6 +29,8 @@ _systemd_dissect() {
     local cur=${COMP_WORDS[COMP_CWORD]} prev_1=${COMP_WORDS[COMP_CWORD-1]} prev_2=${COMP_WORDS[COMP_CWORD-2]} words cword
     local -A OPTS=(
         [STANDALONE]='-h --help --version
+                     --user
+                     --system
                      --discover
                      --no-pager
                      --no-legend
```

**Comment by poettering** [comment]
> Would it be possible to have some validation for the nspawn side of things?

I added a simple CI test for that now.

File: man/systemd-dissect.xml [hunk 2]
```diff
@@ -503,6 +503,17 @@
         <xi:include href="version-info.xml" xpointer="v254"/></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--system</option></term>
+        <term><option>--user</option></term>
+
+        <listitem><para>When used together with <option>--discover</option> controls whether to search for
+        images installed system-wide or in the user's directories in <varname>$HOME</varname>. If neither
+        switch is specified will search within both scopes.</para>
```
Comment (yuwata) [inline man/systemd-dissect.xml]: Maybe better to insert a comma before `will` ?

File: src/import/importd.c [hunk 3]
```diff
@@ -111,6 +111,8 @@ struct Manager {
 
         bool use_btrfs_subvol;
         bool use_btrfs_quota;
+
+        RuntimeScope runtime_scope; /* for now: always RUNTIME_SCOPE_SYSTEM */
```
Comment (yuwata) [inline src/import/importd.c:115]: Maybe better to set `RUNTIME_SCOPE_SYSTEM` explicitly in `manager_new()`, like others do.

**Review by yuwata — APPROVED** [review]
LGTM. Minor suggestions.

File: test/units/TEST-13-NSPAWN.nspawn.sh [hunk 4]
```diff
@@ -1129,7 +1129,7 @@ testcase_unpriv() {
 
     local tmpdir name
     tmpdir="$(mktemp -d /var/tmp/TEST-13-NSPAWN.unpriv.XXX)"
-    name="unpriv-${tmpdir##*.}"
+    name="unprv-${tmpdir##*.}"
```
Comment (bluca) [inline test/units/TEST-13-NSPAWN.nspawn.sh:1132]: Why is this changed?

Comment (poettering) [reply test/units/TEST-13-NSPAWN.nspawn.sh:1132]: to make it shorter. the userns is named after that. #35685 has a patch that removes the userns naming limit, but that's another story...

File: src/dissect/dissect.c [hunk 5]
```diff
@@ -531,14 +538,21 @@ static int parse_argv(int argc, char *argv[]) {
...
 
                         arg_action = ACTION_MAKE_ARCHIVE;
                         break;
 
+                case ARG_SYSTEM:
+                        arg_runtime_scope = (arg_runtime_scope == RUNTIME_SCOPE_USER) ? _RUNTIME_SCOPE_INVALID : RUNTIME_SCOPE_SYSTEM;
+                        break;
+
+                case ARG_USER:
+                        arg_runtime_scope = (arg_runtime_scope == RUNTIME_SCOPE_SYSTEM) ? _RUNTIME_SCOPE_INVALID : RUNTIME_SCOPE_USER;
```
Comment (YHNdnzj) [inline src/dissect/dissect.c]: Hmm, this means --user and --system aren't idempotent? Specifying both but different times could change the behavior? Maybe should just store as flag instead.

**Review by bluca — APPROVED** [review]

File: src/import/export.c [hunk 6]
```diff
@@ -25,6 +25,7 @@
 
 static ImportCompressType arg_compress = IMPORT_COMPRESS_UNKNOWN;
 static ImageClass arg_class = IMAGE_MACHINE;
+static RuntimeScope arg_runtime_scope = _RUNTIME_SCOPE_INVALID;
```
Comment (YHNdnzj) [inline src/import/export.c:28]: We don't usually use arg_ as placeholders? If it's not hooked up I'd just pass _INVALID directly to functions to minimize confusion (here and everywhere)

Comment (poettering) [reply src/import/export.c:28]: as mentioned in the git commit, this is supposed to be hooked up sooner rather than later. i didn't want to add more stuff to the PR that ports importd fully over

but it seemed nicer to me to prep the groundwork here so that we only have to add the cmdline parsers later.

File: src/dissect/dissect.c (hunk 5 above)
Comment (poettering) [reply src/dissect/dissect.c]: not sure i follow?

File: src/import/export.c (hunk 6 above)
Comment (yuwata) [reply src/import/export.c:28]: @YHNdnzj `_RUNTIME_SCOPE_INVALID` means both system and user.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/dissect/dissect.c (hunk 5 above)
Comment (YHNdnzj) [reply src/dissect/dissect.c]: E.g. --system --user --user means user scope is selected eventually, not both

Comment (yuwata) [reply src/dissect/dissect.c]: Hm, interesting...

File: src/import/export.c (hunk 6 above)
Comment (YHNdnzj) [reply src/import/export.c:28]: > as mentioned in the git commit, this is supposed to be hooked up sooner rather than later. i didn't want to add more stuff to the PR that ports importd fully over
> 
> but it seemed nicer to me to prep the groundwork here so that we only have to add the cmdline parsers later.

Maybe add a TODO then?

Comment (poettering) [reply src/import/export.c:28]: added some entries to TODO now
